---
title: "Testat 2"
output:
  word_document: default
  pdf_document: default
---

# Aufgabe 1

Jaaaaaaaaan

# Aufgabe 2

A 2.1-2.2

```{r}
source("dynproLösung.R")
```

```{r}
#st = matrix( c('-','C','G','A','T','C','C','T','G','T',
#              '-','C','A','T','C','G','C','C','T','T'),ncol = 10, byrow = TRUE)
input1 = "CGATCCTGT"
input2 = "CATCGCCTT"

dnA = paste("-", input1, sep = "")
dnB = paste("-", input2, sep = "")

dna1 = strsplit(dnA, split = "")[[1]]
dna2 = strsplit(dnB, split = "")[[1]]

#st = matrix( c(dna1,dna2),ncol = 10, byrow = TRUE)
```



```{r}
# only to get a pretty Alignment output
dynpro.printAlignment = function(decisionPath, st) {
  pointA = 2
  pointB = 2
  alignment = matrix(0L, nrow = 3, ncol = (length(decisionPath) - 1))
  
  for (i in 2:(length(decisionPath))) {
    j = i-1
    d = decisionPath[i]
    if (d == 2) {
      alignment[1, j] = dna1[pointA]
      if (dna1[pointA] == dna2[pointB]) {
        alignment[2, j] = '|'
      } else {
        alignment[2, j] = ' '
      }
      alignment[3, j] = dna2[pointB]
      pointA = pointA + 1
      pointB = pointB + 1
    } else if (d == 1) {
      alignment[1, j] = dna1[pointA]
      alignment[2, j] = ' '
      alignment[3, j] = '-'
      pointA = pointA + 1
    } else if (d == 0) {
      alignment[1, j] = '-'
      alignment[2, j] = ' '
      alignment[3, j] = dna2[pointB]
      pointB = pointB + 1
    }
  }
  return(alignment)
}

#for alignment output without col|row headers (only to prettify)
printM = function(m) {
  write.table(
    format(m, justify = "right"),
    row.names = F,
    col.names = F,
    quote = F
  )
}

dynpro.value = function(i, j, d) {
  if (d == -1) {
    # start
    0
  } else if (d == 0 || d == 1) {
    #gap
    -2
  } else if (d == 2) {
    if (dna1[i] == dna2[j]) {
      # is it match||mismatch
      2
    } else {
      -1
    }
  }
}
dynpro.prevstates = function(i, j, d) {
  if (d == -1) {
    #bei start keine prevstates
    c(0)
  } else if (d == 0) {
    c(i, j - 1)
  } else if (d == 1) {
    c(i - 1, j)
  } else if (d == 2) {
    c(i - 1, j - 1)
  }
}

dynpro.decisions = function(i, j) {
  if (i == 1 && j == 1) {
    c(-1) # start
  } else if (i == 1) {
    c(0) # nur von links
  } else if (j == 1) {
    c(1) # nur von oben
  }
  else{
    c(0, 1, 2) #alles  möglich
    #2 is von links oben
  }
}

result = dynpro.matrix(dna1, dna2, dynpro.value, dynpro.prevstates, dynpro.decisions)

result

decisionPath = dynpro.path(result[[2]], dynpro.prevstates, nrow(result[[1]]), ncol(result[[1]]))

"DecisionPath"
decisionPath
"Alignment"
printM(dynpro.printAlignment(decisionPath, st))

```

```{r}
## Dotplots
#s1 = strsplit(input1, split = "")[[1]]
#s2 = strsplit(input2, split = "")[[1]]
#
#dotPlot(s1, s2, col = c("white", "blue"), xaxt="n", yaxt="n", xlab="", ylab = "") # keine ##Achsenbeschriftung
#axis(1, labels = s1, at=1:length(s1), las=1) # Sequenz als Beschriftung
#axis(2, labels = s2, at=1:length(s2), las=2)
#grid(nx = length(s1), ny = length(s2), col = "black")
```


```{r}
#installiern fall es nich funzt

#install.packages("BiocManager")
#BiocManager::install()
#BiocManager::install("Biostrings")

library(Biostrings)
library(seqinr)
```

Ich weiß hier nich was er haben will...
Wenn wir die Score Matrix aus dem Aufgabenblatt nutzen wie verlangt kommt das halt raus ¯\_(ツ)_/¯
und er meinte in der VL das man hier auch die Aminosäuren vergleichen soll ? aber das wäre beknackt weil man dafür ja BLOSOM50 braucht und das is bei der nächsten  ¯\_(ツ)_/¯

```{r}
#https://www.rdocumentation.org/packages/Biostrings/versions/2.40.2/topics/pairwiseAlignment

s1 = input1
s2 = input2
aa2 = "KIQYKREPNIPSVSLINSLFAWEIRDRI"
aa1 = "KAQYRRECMIFVWEINRL"
mat <- nucleotideSubstitutionMatrix(match = 1, mismatch = -1, baseOnly = TRUE)
#gap opening/extension nicht in der Aufgabe erwähnt und wenn man die ScoreMatrix in der #Aufgabenstellung nimmt wäre es das selbe .... . . oder ich bin zu blöd 
globalAlign <-
  pairwiseAlignment(s1, s2, substitutionMatrix = mat,
                    gapOpening = -2, gapExtension = -2)
globalAlign
localAlign <-
  pairwiseAlignment(s1, s2, type = "local", substitutionMatrix = mat,
                    gapOpening = -2, gapExtension = -2)
localAlign

globalAlign <-
  pairwiseAlignment(s2, s1, substitutionMatrix = mat,
                    gapOpening = -2, gapExtension = -2)
globalAlign
localAlign <-
  pairwiseAlignment(s2, s1, type = "local", substitutionMatrix = mat,
                    gapOpening = -2, gapExtension = -2)
localAlign


```
Hier kommt eig immer KAQYRRE------CMI---FVWEI-NRL raus bei jeglichen Kombinationen

gapOpening kann man leider nicht negativ machen ... also kann man das nicht erzwingen :(
wäre lustig genauso wie extensions


```{R}
## Amino acid global alignment
pa = pairwiseAlignment(aa1, aa2, type = "local",
                  substitutionMatrix = "BLOSUM62",
                  gapOpening = 0, gapExtension = 0)
pa
pa = pairwiseAlignment(aa2, aa1, type = "local",
                  substitutionMatrix = "BLOSUM62",
                  gapOpening = 0, gapExtension = 0)
pa

pa = pairwiseAlignment(aa1, aa2, type = "global",
                  substitutionMatrix = "BLOSUM62",
                  gapOpening = 0, gapExtension = 0)
pa
pa = pairwiseAlignment(aa2, aa1, type = "global",
                  substitutionMatrix = "BLOSUM62",
                  gapOpening = 0, gapExtension = 0)
pa

# for(i in 0:12){
#   for(j in 0:8){
#     p = pairwiseAlignment(aa1, aa2,
#                   substitutionMatrix = "BLOSUM50",
#                   gapOpening = i, gapExtension = j)
#     print(c("Open",i,"ext",j))
#     print(pattern(p))
#     print(subject(p))
#     pp = c(pp,subject(p))
#   }
# }

```
# Aufgabe 3

## Aufgabe 3.1
Sie befinden sich auf Chromosom 12 in Gen NR1H4

## Aufgabe 3.2

Die Aminosäuresequenz wird durch das Gen IFNL (1-3) codiert.